node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/RetiredExpirer.java

```diff
@@ -62,27 +61,32 @@ protected double maintain() {
 
                 NodeList nodes = kv.getValue();
                 attempts++;
-                try (MaintenanceDeployment deployment = new MaintenanceDeployment(application, deployer, metric, nodeRepository())) {
-                    if (!deployment.isValid()) {
-                        log.info("Skipping invalid deployment for " + application);
-                        continue;
-                    }
-
-                    nodeRepository().nodes().setRemovable(application, nodes.asList(), removal.isReusable());
-                    Optional<Long> session = deployment.activate();
-                    String nodeList = String.join(", ", nodes.mapToList(Node::hostname));
-                    if (session.isEmpty()) {
-                        log.info("Failed to redeploy " + application);
-                        continue;
-                    }
-                    log.info("Redeployed " + application + " at session " + session.get() + " to deactivate retired nodes: " + nodeList);
+                if (removeNodes(application, removal, nodes)) {
                     successes++;
                 }
             }
         }
         return attempts == 0 ? 1.0 : ((double)successes / attempts);
     }
 
+    /** Mark nodes as removable and redeploy application */
```

Comment:
I think there was missing locking here since the node list determining what to remove was determined before locking ... and the new code still has that problem.

Reply:
True, will fix.